File: lib/haveapi/client/resource_instance.rb
module HaveAPI::Client
  # Instance of an object from the API.
  # An instance of this class may be in three states:
  # - resolved/persistent - the instance was created by an action that retrieved
  #   it from the API.
  # - unresolved - this instance is an attribute of another instance that was resolved
  #   and will be resolved when first accessed.
  # - not persistent - created by Resource.new, the object was not yet sent to the API.
  class ResourceInstance < Resource
    def initialize(client, api, resource, action: nil, response: nil,
                   resolved: false, meta: nil, persistent: true)
      super(client, api, resource._name)

      @action = action
      @resource = resource
      @resolved = resolved
      @meta = meta
      @persistent = persistent
      @resource_instances = {}

      if response
        if response.is_a?(Hash)
          @params = response
          @prepared_args = response[:_meta][:url_params]

        else
          @response = response
          @params = response.response
          @prepared_args = response.meta[:url_params]
        end

        setup_from_clone(resource)
        define_attributes
      end

      unless @persistent
        setup_from_clone(resource)
        define_implicit_attributes
        define_attributes(@action.input_params)
      end
    end

    def new
      raise NoMethodError.new
    end

    # Invoke +create+ action if the object is not persistent,
    # +update+ action if it is.
    def save
      if @persistent
        method(:update).call

      else
        @action.provide_args
        @response = Response.new(@action, @action.execute(attributes_for_api(@action)))

        if @response.ok?
          @params = @response.response
          define_attributes

        else
          return nil
        end

        @persistent = true
        self
      end
    end

    # Call #save and raise ActionFailed if it fails.
    def save!
      raise ActionFailed.new(@response) if save.nil?
      self
    end

    # Resolve the object (fetch it from the API) if it is not resolved yet.
    def resolve
      return self if @resolved

      @action.provide_args(*@meta[:url_params])
      @response = Response.new(@action, @action.execute({}))
      @params = @response.response

      setup_from_clone(@resource)
      define_attributes

      @resolved = true
      self
    end

    # Return Response object which created this instance.
    def api_response
      @response
    end

    # Return a hash of all object attributes retrieved from the API.
    def attributes
      @params
    end

    def to_s
      "<#{self.class.to_s}:#{object_id}:#{@resource._name}>"
    end

    protected
    # Define access/writer methods for object attributes.
    def define_attributes(params = nil)
      (params || @action.params).each do |name, param|
        case param[:type]
          when 'Resource'
            @resource_instances[name] = find_association(param, @params[name])

            # id reader
            ensure_method(:"#{name}_id") { @params[name][ param[:value_id].to_sym ] }

            # id writer
            ensure_method(:"#{name}_id=") { |id| @params[name][ param[:value_id].to_sym ] = id }

            # value reader
            ensure_method(name) do
              @resource_instances[name] && @resource_instances[name].resolve
            end

            # value writer
            ensure_method(:"#{name}=") do |obj|
              @params[name][ param[:value_id].to_sym ] = obj.method(param[:value_id]).call
              @params[name][ param[:value_label].to_sym ] = obj.method(param[:value_label]).call

              @resource_instances[name] = obj
            end

          else
            # reader
            ensure_method(name) { @params[name] }

            # writer
            ensure_method(:"#{name}=") { |new_val| @params[name] = new_val }
        end
      end
    end

    # Define method +name+ with +block+ if it isn't defined yet.
    def ensure_method(name, &block)
      define_singleton_method(name, &block) unless respond_to?(name)
    end

    # Define nil references to resource attributes.
    # Used only for not-persistent objects.
    def define_implicit_attributes
      @params = {}

      @action.input_params.each do |name, param|
        @params[name] = {} if param[:type] == 'Resource'
      end
    end

    # Return a hash of all attributes suitable to be sent to the API +action+.
    def attributes_for_api(action)
      ret = {}

      return ret if action.input_layout != :object

      action.input_params.each do |name, param|
        case param[:type]
          when 'Resource'
            ret[name] = @params[name][ param[:value_id].to_sym ]

          else
            ret[name] = @params[name]
        end
      end

      ret
    end

    # Find associated resource and create its unresolved instance.
    def find_association(res_desc, res_val)
      return nil unless res_val

      tmp = @client

      res_desc[:resource].each do |r|
        tmp = tmp.method(r).call
      end

      # FIXME: read _meta namespace from description
      ResourceInstance.new(
          @client,
          @api,
          tmp,
          action: tmp.actions[:show],
          resolved: res_val[:_meta][:resolved],
          response: res_val[:_meta][:resolved] ? res_val : nil,
          meta: res_val[:_meta]
      )
    end

    # Override Resource.default_action_input_params.
    def default_action_input_params(action)
      attributes_for_api(action)
    end
  end
end

File: clients/js/src/haveapi/resource_instance.js
/**
 * @class ResourceInstance
 * @classdesc Represents an instance of a resource from the API. Attributes
 *            are accessible as properties. Associations are directly accessible.
 * @param {HaveAPI.Client}          client
 * @param {HaveAPI.Client.Action}   action    Action that created this instance.
 * @param {HaveAPI.Client.Response} response  If not provided, the instance is either
 *                                            not resoved or not persistent.
 * @param {Boolean}                 shell     If true, the resource is just a shell,
 *                                            it is to be fetched from the API. Used
 *                                            when accessed as an association from another
 *                                            resource instance.
 * @param {Boolean}                 item      When true, this object was returned in a list,
 *                                            therefore response is not a Response instance,
 *                                            but just an object with parameters.
 * @memberof HaveAPI.Client
 */
function ResourceInstance (client, parent, action, response, shell, item) {
	this._private = {
		client: client,
		parent: parent,
		action: action,
		response: response,
		name: action.resource._private.name,
		description: action.resource._private.description
	};

	if (!response) {
		if (shell !== undefined && shell) { // association that is to be fetched
			this._private.resolved = false;
			this._private.persistent = true;

			var that = this;

			action.directInvoke(function(c, response) {
				that.attachResources(that._private.action.resource._private.description, response.meta().url_params);
				that.attachActions(that._private.action.resource._private.description, response.meta().url_params);
				that.attachAttributes(response.response());

				that._private.resolved = true;

				if (that._private.resolveCallbacks !== undefined) {
					for (var i = 0; i < that._private.resolveCallbacks.length; i++)
						that._private.resolveCallbacks[i](that._private.client, that);

					delete that._private.resolveCallbacks;
				}
			});

		} else { // a new, empty instance
			this._private.resolved = true;
			this._private.persistent = false;

			this.attachResources(this._private.action.resource._private.description, action.providedIdArgs);
			this.attachActions(this._private.action.resource._private.description, action.providedIdArgs);
			this.attachStubAttributes();
		}

	} else if (item || response.isOk()) {
		this._private.resolved = true;
		this._private.persistent = true;

		var metaNs = client.apiSettings.meta.namespace;
		var idArgs = item ? response[metaNs].url_params : response.meta().url_params;

		this.attachResources(this._private.action.resource._private.description, idArgs);
		this.attachActions(this._private.action.resource._private.description, idArgs);
		this.attachAttributes(item ? response : response.response());

	} else {
		// FIXME
	}
};

ResourceInstance.prototype = new BaseResource();

/**
 * @callback HaveAPI.Client.ResourceInstance~resolveCallback
 * @param {HaveAPI.Client} client
 * @param {HaveAPI.Client.ResourceInstance} resource
 */

/**
 * A shortcut to {@link HaveAPI.Client.Response#isOk}.
 * @method HaveAPI.Client.ResourceInstance#isOk
 * @return {Boolean}
 */
ResourceInstance.prototype.isOk = function() {
	return this._private.response.isOk();
};

/**
 * Return the response that this instance is created from.
 * @method HaveAPI.Client.ResourceInstance#apiResponse
 * @return {HaveAPI.Client.Response}
 */
ResourceInstance.prototype.apiResponse = function() {
	return this._private.response;
};

/**
 * Save the instance. It calls either an update or a create action,
 * depending on whether the object is persistent or not.
 * @method HaveAPI.Client.ResourceInstance#save
 * @param {HaveAPI.Client~replyCallback} callback
 */
ResourceInstance.prototype.save = function(callback) {
	var that = this;

	function updateAttrs(attrs) {
		for (var attr in attrs) {
			that._private.attributes[ attr ] = attrs[ attr ];
		}
	};

	function replyCallback(c, reply) {
		that._private.response = reply;
		updateAttrs(reply);

		if (callback !== undefined)
			callback(c, that);
	}

	if (this._private.persistent) {
		this.update.directInvoke(replyCallback);

	} else {
		this.create.directInvoke(function(c, reply) {
			if (reply.isOk())
				that._private.persistent = true;

			replyCallback(c, reply);
		});
	}
};

ResourceInstance.prototype.defaultParams = function(action) {
	ret = {}

	for (var attr in this._private.attributes) {
		var desc = action.description.input.parameters[ attr ];

		if (desc === undefined)
			continue;

		switch (desc.type) {
			case 'Resource':
				ret[ attr ] = this._private.attributes[ attr ][ desc.value_id ];
				break;

			default:
				ret[ attr ] = this._private.attributes[ attr ];
		}
	}

	return ret;
};

/**
 * Resolve an associated resource.
 * A shell {@link HaveAPI.Client.ResourceInstance} instance is created
 * and is fetched asynchronously.
 * @method HaveAPI.Client.ResourceInstance#resolveAssociation
 * @private
 * @return {HaveAPI.Client.ResourceInstance}
 */
ResourceInstance.prototype.resolveAssociation = function(attr, path, url) {
	var tmp = this._private.client;

	for(var i = 0; i < path.length; i++) {
		tmp = tmp[ path[i] ];
	}

	var obj = this._private.attributes[ attr ];
	var metaNs = this._private.client.apiSettings.meta.namespace;
	var action = tmp.show;
	action.provideIdArgs(obj[metaNs].url_params);

	if (obj[metaNs].resolved)
		return new Client.ResourceInstance(
			this._private.client,
			action.resource._private.parent,
			action,
			obj,
			false,
			true
		);

	return new Client.ResourceInstance(
		this._private.client,
		action.resource._private.parent,
		action,
		null,
		true
	);
};

/**
 * Register a callback that will be called then this instance will
 * be fully resolved (fetched from the API).
 * @method HaveAPI.Client.ResourceInstance#whenResolved
 * @param {HaveAPI.Client.ResourceInstance~resolveCallback} callback
 */
ResourceInstance.prototype.whenResolved = function(callback) {
	if (this._private.resolved)
		callback(this._private.client, this);

	else {
		if (this._private.resolveCallbacks === undefined)
			this._private.resolveCallbacks = [];

		this._private.resolveCallbacks.push(callback);
	}
};

/**
 * Attach all attributes as properties.
 * @method HaveAPI.Client.ResourceInstance#attachAttributes
 * @private
 * @param {Object} attrs
 */
ResourceInstance.prototype.attachAttributes = function(attrs) {
	this._private.attributes = attrs;
	this._private.associations = {};

	var metaNs = this._private.client.apiSettings.meta.namespace;

	for (var attr in attrs) {
		if (attr === metaNs)
			continue;

		this.createAttribute(attr, this._private.action.description.output.parameters[ attr ]);
	}
};

/**
 * Attach all attributes as null properties. Used when creating a new, empty instance.
 * @method HaveAPI.Client.ResourceInstance#attachStubAttributes
 * @private
 */
ResourceInstance.prototype.attachStubAttributes = function() {
	var attrs = {};
	var params = this._private.action.description.input.parameters;

	for (var attr in params) {
		switch (params[ attr ].type) {
			case 'Resource':
				attrs[ attr ] = {};
				attrs[ attr ][ params[attr].value_id ] = null;
				attrs[ attr ][ params[attr].value_label ] = null;
				break;

			default:
				attrs[ attr ] = null;
		}
	}

	this.attachAttributes(attrs);
};

/**
 * Define getters and setters for an attribute.
 * @method HaveAPI.Client.ResourceInstance#createhAttribute
 * @private
 * @param {String} attr
 * @param {Object} desc
 */
ResourceInstance.prototype.createAttribute = function(attr, desc) {
	var that = this;

	switch (desc.type) {
		case 'Resource':
			Object.defineProperty(this, attr, {
				get: function() {
						if (that._private.associations.hasOwnProperty(attr))
							return that._private.associations[ attr ];

						return that._private.associations[ attr ] = that.resolveAssociation(
							attr,
							desc.resource,
							that._private.attributes[ attr ].url
						);
					},
				set: function(v) {
						that._private.attributes[ attr ][ desc.value_id ]    = v.id;
						that._private.attributes[ attr ][ desc.value_label ] = v[ desc.value_label ];
					}
			});

			Object.defineProperty(this, attr + '_id', {
				get: function()  { return that._private.attributes[ attr ][ desc.value_id ];  },
				set: function(v) { that._private.attributes[ attr ][ desc.value_id ] = v;     }
			});

			break;

		default:
			Object.defineProperty(this, attr, {
				get: function()  { return that._private.attributes[ attr ];  },
				set: function(v) { that._private.attributes[ attr ] = v;     }
			});
	}
};
